opts.go

```diff
@@ -0,0 +1,156 @@
+type Params map[PName]PValue
+type PName string
+type PValue string
+
+var pNameRegexString = `([A-Za-z_][0-9A-Za-z_]*)`
+var pNameRegex = regexp.MustCompile(`^` + pNameRegexString + `$`)
+var paramRegex = regexp.MustCompile(`{{ ` + pNameRegexString + ` }}`)
+
+func toPName(input string) (PName, error) {
+	// check whether string is valid PName
+	if !pNameRegex.MatchString(input) {
+		return PName(""), fmt.Errorf("Invalid identifer: %s", input)
+	}
+	return PName(input), nil
+}
+
+// specific Params -> String transformation
+func (params *Params) String() string {
+	return fmt.Sprint(*params)
+}
+
+// tell `flag` package how to parse/store param args
+func (params *Params) Set(input string) error {
+	// for each `--param` flag...
+	for _, p := range strings.Split(input, ",") {
+		// split params on ,
+		pElems := strings.SplitN(p, ":", 2)
+		// check number of elements in param arg
+		switch len(pElems) {
+		case 0:
+			return fmt.Errorf("Missing argument to `--param` flag.")
+		case 1:
+			return fmt.Errorf("Missing value in parameter arg: %s", p)
+		}
+		// validate string as PName identifer
+		pName, err := toPName(pElems[0])
+		if err != nil {
+			return err
+		}
+		// add pName -> pVal entry to params map
+		pValue := PValue(pElems[1])
+		(*params)[pName] = pValue
+	}
+	return nil
+}
+
+// Match holds the parameter declaration (e.g. `{{ N }}`) and name (e.g. `N`)
+type Match struct {
+	pDeclaration []byte
+	pName        PName
+}
+
+/* resolve all param names (keys in `params` map) to their resp. values */
+func replaceParams(fileData []byte, params Params) []byte {
+	// get all unique param declarations from test file
+	matches := uniqueParamDeclarations(fileData)
+	for _, match := range matches {
+		// check whether the reference parameter is defined
+		if pVal, ok := params[match.pName]; ok {
+			// parameter found, replace all occurrences with its value
+			fileData = bytes.Replace(fileData, match.pDeclaration, []byte(pVal), -1)
+		} else {
+			// parameter not found, fail
+			fatal(fmt.Sprintf("Error: Parameter %s not specified\n", match.pName))
+		}
+	}
+	return fileData
+}
+
+/* returns all unique parameter declarations in a test file */
+func uniqueParamDeclarations(fileData []byte) []Match {
+    // get all regex results
+	regexResults := paramRegex.FindAllSubmatch(fileData, -1)
+	return getUniqueMatches(regexResults)
+}
+
+func getUniqueMatches(regexResults [][][]byte) []Match {
+	/* there's not a great way to do this in Go... */
+	var unique []Match
+	for _, result := range regexResults {
+		match := Match{
+			pDeclaration: result[0],
+			pName:        PName(result[1]),
+		}
+		unique = appendIfMissing(unique, match)
+	}
+	return unique
+}
+
+func appendIfMissing(matches []Match, match Match) []Match {
+	// check whether `match` already in `matches`, append if not
+	for _, m := range matches {
+		if bytes.Compare(m.pDeclaration, m.pDeclaration) == 0 && m.pName == match.pName {
```

Comment:
see comment above. This would go away. But note here that if p.Declaration is the same, pName should be the same too since it's a submatch after all. (right?)